### PR TITLE
Revise DetectionOutput reference implementation

### PR DESCRIPTION
**Pull request by mateusztabaka**
Ticket: 37433

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 1]
```diff
@@ -278,17 +278,23 @@ class DetectionOutputImpl: public ExtLayerBase {
             }
         }
 
+        const int num_results = outputs[0]->getTensorDesc().getDims()[2];
```
Comment (iefode) [inline inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:281]: @dmitry-gorokhov Please take a look CPU part

File: ngraph/core/src/op/detection_output.cpp [hunk 2]
```diff
@@ -45,16 +45,227 @@ op::DetectionOutput::DetectionOutput(const Output<Node>& box_logits,
...
+
+    std::vector<Dimension> output_shape{1, 1};
+    if (num_images.is_dynamic())
+    {
+        output_shape.push_back(Dimension::dynamic());
     }
     else
     {
-        set_output_type(0, element::f32, PartialShape::dynamic());
+        size_t num_images_val = num_images.get_interval().get_min_val();
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/detection_output.cpp:241]: Why do you use `num_images.get_interval().get_min_val()` here? Why don't use just `num_images`?

File: ngraph/core/src/op/detection_output.cpp [hunk 3]
```diff
@@ -45,16 +45,227 @@ op::DetectionOutput::DetectionOutput(const Output<Node>& box_logits,
...
+        if (m_attrs.keep_top_k[0] > 0)
+        {
+            output_shape.push_back(num_images_val * m_attrs.keep_top_k[0]);
+        }
+        else if (m_attrs.top_k > 0)
+        {
+            output_shape.push_back(num_images_val * m_attrs.top_k * m_attrs.num_classes);
+        }
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/detection_output.cpp:249]: Don't quite understand why the output shape is calculated differently. In the CPU implementation code we have the following code:

```cpp
        auto dst_data_size = N * _keep_top_k * DETECTION_SIZE * sizeof(float);

        if (dst_data_size > outputs[0]->byteSize()) {
            return OUT_OF_BOUNDS;
        }
```
also we have the following code:
```cpp
        if (count < N*_keep_top_k) {
            // marker at end of boxes list
            dst_data[count * DETECTION_SIZE + 0] = -1;
        }
```

So it looks like that the output shape should always be calculated that way.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/python/tests/test_ngraph/test_create_op.py [hunk 4]
```diff
@@ -932,11 +932,11 @@ def test_detection_output(int_dtype, fp_dtype):
         "nms_threshold": fp_dtype(0.645),
     }
 
-    box_logits = ng.parameter([4, 1, 5, 5], fp_dtype, "box_logits")
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply ngraph/python/tests/test_ngraph/test_create_op.py:935]: DetectionOutput's validate_and_infer_types did not have restrictions for input

File: ngraph/core/src/op/detection_output.cpp (hunk 2 above)
Comment (mateusztabaka) [reply ngraph/core/src/op/detection_output.cpp]: right, I simplified that code a little.

File: ngraph/core/src/op/detection_output.cpp (hunk 3 above)
Comment (mateusztabaka) [reply ngraph/core/src/op/detection_output.cpp]: if keep_top_k == -1 then we can't use it to determine the shape, so we have to use other attributes for that.

**Review by lazarevevgeny — APPROVED** [review]
Spec changes as well as op shape infer looks good.

**Review by ggalieroc — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
CPU part looks good to me

**Pull request merged**